### PR TITLE
New SMB storage backend

### LINKED ISSUE #6092: SMB/CIFS password visible in processlist
When using external storage "SMB/CIFS" username and password becomes visible to users logged in on the server.

The password is visible when listing processes on the server (ps -ef)

This is a running process:
sh -c smbclient -N -U 'DOMAIN/UID%PASSWORD' -O 'TCP_NODELAY IPTOS_LOWDELAY SO_KEEPALIVE SO_RCVBUF=8192 SO_SNDBUF=8192' -O 'TCP_NODELAY IPTOS_LOWDELAY SO_KEEPALIVE SO_RCVBUF=8192 SO_SNDBUF=8192' -d 0 '//sambaserver/share' -c 'dir "UID\Documents\Download\desktop.ini"' 2>/dev/null

### LINKED ISSUE #9467: SMB/CIFS password visible in owncloud.log
based on discussion in https://github.com/owncloud/core/issues/6092#issuecomment-48082490:

Owncloud logs the password when using smbclient. I have in my logfiles:

    {"reqId":"53b7591f67f7a","app":"PHP","message":"popen(TZ=UTC smbclient -N  -U 'user%TopSecretPassword' -O 'TCP_NODELAY IPTOS_LOWDELAY SO_KEEPALIVE SO_RCVBUF=8192 SO_SNDBUF=8192' ...

### LINKED ISSUE #9480: Use native library for SMB mounts
Try using something that binds again libsmbclient or http://sourceforge.net/projects/phpsambaext/ instead of using the `smbclient` command line client.

As discussed with @Xenopathic 

CC @DeepDiver1975 @icewind1991

### PR DESCRIPTION
**Pull request by icewind1991**
New SMB storage backend based on https://github.com/icewind1991/SMB which supports either wrapping smbclient or use libsmbclient directly trough libsmbclient-php.

All tests pass locally (both with smbclient and libsmbclient)

Both methods also provide a nice speedup compared to smb4php, on my local machine I get the following times for running the smb unit tests

| old (smb4php) | new (smbclient) | new (libsmbclient) |
| --- | --- | --- |
| 16s | 8s | 1s |

cc @DeepDiver1975 @PVince81 @Xenopathic 

Fixes #6092, fixes #9467, fixes #9480

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Should be revived for 8.1

**Comment by DeepDiver1975** [comment]
> Should be revived for 8.1

indeed

**Comment by PVince81** [comment]
Please rebase, if it makes sense at this point

**Comment by icewind1991** [comment]
@PVince81 @DeepDiver1975 can we merge this once stable8 is branched off?

**Comment by PVince81** [comment]
So, do the unit tests pass now ?

File: apps/files_external/lib/smb.php [hunk 1]
```diff
@@ -8,139 +8,271 @@
...
 	}
 
 	/**
 	 * check if smbclient is installed
 	 */
 	public static function checkDependencies() {
-		$smbClientExists = (bool) \OC_Helper::findBinaryPath('smbclient');
-		return $smbClientExists ? true : array('smbclient');
+		if (function_exists('shell_exec')) {
+			$output = shell_exec('command -v smbclient 2> /dev/null');
```
Comment (LukasReschke) [inline apps/files_external/lib/smb.php]: The left code uses `OC_Helper::findBinaryPath` which looks like the more modern way to do this?

File: apps/files_external/appinfo/app.php [hunk 2]
```diff
@@ -21,6 +21,8 @@
 OC::$CLASSPATH['OC_Mount_Config'] = 'files_external/lib/config.php';
 OC::$CLASSPATH['OCA\Files\External\Api'] = 'files_external/lib/api.php';
 
+require_once __DIR__ . '/../3rdparty/autoload.php';
```
Comment (MorrisJobke) [inline apps/files_external/appinfo/app.php]: Is this really needed?

Comment (icewind1991) [inline apps/files_external/appinfo/app.php]: Yes, it uses the composer autoloader

**Comment by icewind1991** [comment]
> So, do the unit tests pass now ?

We have passing unit tests for this over at bhttps://github.com/owncloud/core/issues/13922

File: apps/files_external/lib/smb.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_external/lib/smb.php]: A problem with old branches and lazy conflict resolving, fixed

**Comment by PVince81** [comment]
CC @jnfrmarks once this is merged we need to regression test SMB mounts

**Comment by DeepDiver1975** [comment]
I'd like to see unit tests passing against a windows share. Once done we can merge this .... finally

**Comment by PVince81** [comment]
- [ ] BUG: enable encryption, create empty file in the web UI, download that file :no_entry_sign: FAIL

**Comment by PVince81** [comment]
- [ ] BUG: delete file when encryption is enabled:
```
{"reqId":"e61465d9445dfaee825f73d4669f06e9","remoteAddr":"127.0.0.1","app":"index","message":"Exception: {\"Message\":\"Invalid request for smb:\\\/\\\/localhost\\\/vincent\\\/temp\\\/smbmount\\\/P1050409.JPG\",\"Code\":2,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/smb\\\/src\\\/NativeState.php(72): Icewind\\\\SMB\\\\NativeState->handleError('smb:\\\/\\\/localhost...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/smb\\\/src\\\/NativeState.php(146): Icewind\\\\SMB\\\\NativeState->testResult(false, 'smb:\\\/\\\/localhost...')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/smb\\\/src\\\/NativeShare.php(152): Icewind\\\\SMB\\\\NativeState->unlink('smb:\\\/\\\/localhost...')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/smb.php(122): Icewind\\\\SMB\\\\NativeShare->del('\\\/temp\\\/smbmount\\\/...')\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_trashbin\\\/lib\\\/storage.php(84): OC\\\\Files\\\\Storage\\\\SMB->unlink('P1050409.JPG')\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(806): OCA\\\\Files_Trashbin\\\\Storage->unlink('P1050409.JPG')\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(442): OC\\\\Files\\\\View->basicOperation('unlink', '\\\/folder\\\/P105040...', Array)\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(651): OC\\\\Files\\\\View->unlink('\\\/folder\\\/P105040...')\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/ajax\\\/delete.php(31): OC\\\\Files\\\\Filesystem::unlink('\\\/folder\\\/P105040...')\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(135) : runtime-created function(1): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#10 [internal function]: __lambda_func(Array)\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(250): call_user_func('\\\\x00lambda_3371', Array)\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/base.php(763): OC\\\\Route\\\\Router->match('\\\/apps\\\/files\\\/aja...')\\n#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/index.php(36): OC::handleRequest()\\n#14 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/smb\\\/src\\\/NativeState.php\",\"Line\":44}","level":4,"time":"2015-02-09T10:26:56+00:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/delete.php"}
```

**Comment by icewind1991** [comment]
@PVince81 fixed both issues, thanks for testing

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
> I'd like to see unit tests passing against a windows share. Once done we can merge this .... finally

Ran the unit tests against a windows server 2012R2 vm

**Comment by DeepDiver1975** [comment]
> Ran the unit tests against a windows server 2012R2 vm

great - thx

**Comment by PVince81** [comment]
@icewind1991 if I remember well you said there were actually two implementations: one used when libsmbclient is available and one that falls back to smbclient. Is that still the case ?
Did you run the test for both cases ?

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
> Did you run the test for both cases ?

Did now, found an upstream [bug](https://bugzilla.samba.org/show_bug.cgi?id=11094) and worked around it. Tests now run for me against both windows and linux servers with both backends

**Comment by DeepDiver1975** [comment]
@MorrisJobke we have to respect the different operation modes in our autotest execution. Might be tricky .....

**Comment by MorrisJobke** [comment]
> @MorrisJobke we have to respect the different operation modes in our autotest execution. Might be tricky .....

What does that mean?

**Comment by icewind1991** [comment]
Testing both the libsmbclient and the smblcient implementations of the smb library

**Comment by MorrisJobke** [comment]
> Testing both the libsmbclient and the smblcient implementations of the smb library

How to do that?

**Comment by icewind1991** [comment]
enabling disabling the libsmbclient extension

**Comment by DeepDiver1975** [comment]
I guess we should follow the mechanism in 3rdparty and remove the tests folder from icewind/smb and icewind/streams

@icewind1991 or maybe even you don't package test within the composer package?

**Comment by DeepDiver1975** [comment]
come on @owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
rebased and tested against samba server running inside docker :+1: 

issues with windows are to be seen as local network/setup issues

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
@PVince81 @MorrisJobke can this be merged?

**Comment by MorrisJobke** [comment]
Yes. Tested :+1:

**Comment by DeepDiver1975** [comment]
kungfupanda2

Great job! @icewind1991

**Comment by LukasReschke** [comment]
Please add as feature to https://github.com/owncloud/core/wiki/ownCloud-8.1-Features as well – THX.

**Comment by DeepDiver1975** [comment]
> Please add as feature to https://github.com/owncloud/core/wiki/ownCloud-8.1-Features as well – THX.

done - thx for the reminder

**Comment by derkostka** [comment]
My Questions: smbclient is caching the file to download in /tmp, which may be a problem for larger files. Is this still the case for libsmbclient ? Is the cache Folder configurable somewhere ?

Currently, i am downloading a 2,15 GB File, /tmp is 1024 MB only and 100% filled right now. Will the download work or will it break ?

Thanks for the Info !

**Comment by RealRancor** [comment]
Ref for the documentation of this new "libsmbclient" / "libsmbclient-php" dependency:

https://github.com/owncloud/documentation/issues/1447

It could be really useful to document something like this before the release of the version including it.

**Pull request merged**